### ISSUE TITLE
Rework underlying API for encoding/decoding

### ISSUE THREAD

**Issue by neoeinstein** [xyncro/chiron#72]
The current primary interface for using _Chiron_ is based around functions which use statically-resolved type parameters (SRTP). This API should remain in place, but it should instead be based on an API that doesn't rely on SRTP.

- [ ] Add symmetric encoder and decoder types:
 - `type Decoder<'a> = Json -> JsonResult<'a>`
 - `type Encoder<'a> = 'a -> Json`
- [ ] Extract the default encoder/decoder functions from `ToJsonDefaults` and `FromJsonDefaults` into functions under `Encode` and `Decode` modules, respectively
- [ ] Add monadic and applicative functions for `Decoder<'a>`
- [ ] Add functor and chaining functions for `Encoder<'a>`
- [ ] Create a `fromJson {}` computation expression, adapted from `json {}`

**Comment by kolektiv** [comment]
If we're going to have a fromJson CE (not a bad idea) would it make sense to introduce a toJson CE as well? The json CE would probably remain as a more "general" CE perhaps...

**Comment by neoeinstein** [comment]
Yeah. I've just been re-looking at the `Encoder<'a>` type and thinking that it should be a state monad instead of just the function it is now, so as to support "mixin" encoders. If so, then that leads to a `toJson {}` CE to support that. We could still have the `json {}` CE supported. Though I'm of two minds on retiring it, I'll default to keeping it supported.

**Comment by kolektiv** [comment]
The `json {}` CE does support more than just mapping to/from. With the `Json.Optic.*` support, etc. it supports querying and modification of JSON structures. There is an argument as to whether that's needed outside of a mapping context, but it does so at the moment...

**Comment by neoeinstein** [comment]
Right. I can see that. Maybe I just haven't had a use case for doing reading during a `ToJson` or writing during a `FromJson`. The `Json.Optic.*` support is definitely still going to be useful. This may be something that I need to throw against the wall to really see what it looks like.

**Comment by neoeinstein** [comment]
Ok, I can see a reason for the `json {}` CE. It exists as a composition of the "state" and "either" monads. What we really need is to "run" the CE. The CE needs to be extended so that it can bind to `Encoder`s and `Decoder`s. That would allow the more direct serialization/deserialization method to live along side the more general-purpose `json {}` method.
